### PR TITLE
Add NpDrmPackage

### PR DESCRIPTION
**Pull request by devnoname120**
Reversed and documented by @173210.
See http://173210.github.io/scenpdrmpackage

I'm not sure about the rule for prefixing function names. Should I remove the `_`?

### CODE REVIEW

**Comment by yifanlu** [comment]
Yeah, remove the "_" if there's colliding names, the convention is to add "ForKernel" or "ForDriver" to ones in kernel.

**Comment by 173210** [comment]
In my opinion:
> I think it is legitimate because they are intended for the internal use. Those functions are actually primitive and a userland static library gives more advanced functions. Probably they are expected to be called just by the library and users will use the library instead of calling them by themselves.

Note that the library is on userland but not on kernel. Anyway, I don't think the prefix really matters in the actual use.

**Comment by d3m3vilurr** [comment]
generically, `_` mean `private` or `internal function`.
so i also agree yifanlu`s opinion.
how about just remove `_`?

**Comment by devnoname120** [comment]
Many functions from the pastebin file seem to lose one `_` when added to `db.json`. If they are prefixed with `__`, then it becomes `_`, and `_` becomes nothing. But this is not the case for all of them.

@yifanlu Some rules should be defined. There are currently many discepancies in `db.json`.

**Comment by 173210** [comment]
I think that is because those names are cited from the binary of Sony. However, some are determined in the community and they cannot be distinguished from the "real" names. It is no longer possible and doesn't make sense to track real names. We should have our own rules.

**Comment by devnoname120** [comment]
@yifanlu Let's merge it as is, it will be fixed with the rest once the standard is defined.

**Comment by devnoname120** [comment]
This pull request is broken since the changes have been overwritten by others. I'll try to recover the changes.

**Comment by devnoname120** [comment]
Done.

File: include/psp2/npdrmpackage.h [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+ * Read the header of the PKG and initialize the context
+ *
+ * @param buffer - The buffer containing the header of PKG.
+ * @param size - The size of buffer. The minimum confirmed value is 0x8000.
+ * @param zero - Unknown. Supposed to be set to 0.
+ * @param identifier - arbitrary value [0, 6) ORed with (1 << 8) or 0.
+ *                     If it is set to 0, the function just checks the header
+ *                     and doesn't create the context.
+ */
+int _sceNpDrmPackageCheck(const void *buffer, SceSize size, int zero, unsigned int identifier);
```
Comment (DaveeFTW) [inline include/psp2/npdrmpackage.h]: Can you add documentation for the return value?

File: include/psp2/npdrmpackage.h [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+/**
+ * Decrypt a PKG
+ *
+ * @param buffer - The buffer containing the content of the PKG.
+ * @param size - The size of the buffer. The minimum confirmed value is 0x20.
+ * @param opt - The options.
+ */
+int _sceNpDrmPackageDecrypt(void * restrict buffer, SceSize size, _sceNpDrmPackageDecrypt_opt * restrict opt);
```
Comment (DaveeFTW) [inline include/psp2/npdrmpackage.h]: same here

File: include/psp2/npdrmpackage.h [hunk 3]
```diff
@@ -0,0 +1,58 @@
+/**
+ * \file
+ * \brief Header file which defines NpDrmPackage related variables and functions
+ *
+ * Copyright (C) 2015 PSP2SDK Project
```
Comment (DaveeFTW) [inline include/psp2/npdrmpackage.h]: do you mean 2015?

File: db.json [hunk 4]
```diff
@@ -3552,6 +3552,20 @@
...
+                "functions": {
+                    "_sceNpDrmPackageCheck": 2715321850,
+                    "_sceNpDrmPackageDecrypt": 3606076108
+                },
+                "kernel": false,
+                "nid": 2287029682,
+                "variables": {}
+            }
+        },
+        "nid": 120
```
Comment (DaveeFTW) [inline db.json]: need module nid

File: include/psp2/npdrmpackage.h (hunk 3 above)
Comment (devnoname120) [inline include/psp2/npdrmpackage.h]: Is this date thing really necessary? It seems to be an unnecessary burden, although minor.

File: include/psp2/npdrmpackage.h (hunk 1 above)
Comment (devnoname120) [inline include/psp2/npdrmpackage.h]: I suppose it's < 0 on error, otherwise 0. Need @173210 's confirmation.

File: db.json (hunk 4 above)
Comment (DaveeFTW) [inline db.json]: Most do, yes. We have the technology to fill these in properly now.

File: include/psp2/npdrmpackage.h (hunk 3 above)
Comment (173210) [inline include/psp2/npdrmpackage.h]: We should follow the existing code anyway.

File: include/psp2/npdrmpackage.h (hunk 1 above)
Comment (173210) [inline include/psp2/npdrmpackage.h]: != 0 on error, otherwise 0.

File: db.json (hunk 4 above)
Comment (devnoname120) [inline db.json]: I think that this should be done with the massive overhault, including the _ change.

**Comment by devnoname120** [comment]
Done the changes except `_`.

File: include/psp2/npdrmpackage.h (hunk 3 above)
Comment (devnoname120) [inline include/psp2/npdrmpackage.h]: OK.

**Pull request merged**